### PR TITLE
added tests for Host::Base#merge_facts and fixed a few bugs along the way

### PR DESCRIPTION
**Pull request by dmitri-d**

### CODE REVIEW

File: app/models/host/base.rb [hunk 1]
```diff
@@ -54,46 +55,46 @@ def importFacts name, facts
...
+        fact_values.build(values.collect { |v| {:value => v, :fact_name_id => fact_name_id} })
+      end
+    end
+
+    def normalize_facts(facts)
+      facts.inject({}) do |hash, fact|
+         name, values = *fact
+
+         values = values.is_a?(Array) ? values : [values]
+         values = values.reject { |v| v.nil? }.collect { |v| v.to_s }
```
Comment (ohadlevy) [inline app/models/host/base.rb:95]: v.present? or blank maybe instead?

Comment (ohadlevy) [inline app/models/host/base.rb:95]: btw: I'm not sure it makes sense to run two loops here

File: app/models/host/base.rb [hunk 2]
```diff
@@ -54,46 +55,46 @@ def importFacts name, facts
...
+        (Set.new(values.collect(&:value)) <=> Set.new(normalized_facts[name])) == 0 ?
+          to_delete :
+          (to_delete << values.collect(&:id)).flatten
       end
-
       FactValue.delete(deletions) unless deletions.empty?
 
-      # Get FactNames in one call
       fact_names = FactName.maximum(:id, :group => 'name')
+      db_facts = fact_values.includes(:fact_name).index_by(&:name)
```
Comment (ohadlevy) [inline app/models/host/base.rb:81]: why query the db again?

Comment (dmitri-d) [inline app/models/host/base.rb:81]: because the original might have changed.

File: app/models/host/base.rb (hunk 1 above)
Comment (dmitri-d) [inline app/models/host/base.rb:95]: not sure I follow. As opposed to?

**Pull request closed without merging**